File: 5-longest-palindromic-substring/5-longest-palindromic-substring.cpp
class Solution {
private:
  static string prepareString(string s) {
    vector<char> buffer;
    buffer.push_back('^');
    for (char ch: s) {
      buffer.push_back('#');
      buffer.push_back(ch);
    }
    buffer.push_back('#');
    buffer.push_back('$');
    string result(buffer.begin(), buffer.end());
    return result;
  }
public:
  string longestPalindrome(string s) {
    string str = prepareString(s);
    auto len = str.length();
    int index = 0;
    int center = 0;
    int right = 0;
    vector<int> dp(len, 1);
    for (int i = 1; i < len - 1 - dp[index]; i++) {
      if (i < right - 1) {
        dp[i] = min(right - i, dp[2*center - i]);
      }
      while (str[i - dp[i]] == str[i + dp[i]]) {
        dp[i] += 1;
      }
      if (i + dp[i] > right) {
        center = i;
        right = i + dp[i];
      }
      if (dp[i] > dp[index]) {
        index = i;
      }
    }
    return s.substr((index-dp[index])/2, dp[index] - 1);
  }
};

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
  const str = ['^', ...([].concat(...([...s].map(ch => ['#', ch])))), '#', '$']
  const len = str.length
  const dp = new Array(len).fill(1)
  let c = 0
  let r = 0
  let index = 1
  for (let i = 1; i < len - 1; i++) {
    if (i < (r - 1)) {
      dp[i] = Math.min(dp[2*c-i], r-i)
    }
    while (str[i-dp[i]] === str[i+dp[i]]) {
      dp[i] += 1
    }
    if ((i + dp[i]) > r) {
      c = i
      r = c + dp[i]
    }
    if (dp[i] > dp[index]) {
      index = i
    }
  }
  return s.substr(Math.floor((index-dp[index])/2), dp[index]-1)
};